eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelTypes.scala

```diff
@@ -145,11 +196,51 @@ final case class RES_GETINFO(nodeId: PublicKey, channelId: ByteVector32, state:
        88      `8b 8888888888  "Y8888P"  88            `"Y8888Y"'   88      `888  "Y8888P"  8888888888  "Y8888P"
  */
 
-sealed trait ChannelCommandResponse
-object ChannelCommandResponse {
-  case object Ok extends ChannelCommandResponse { override def toString  = "ok" }
-  case class ChannelOpened(channelId: ByteVector32) extends ChannelCommandResponse { override def toString  = s"created channel $channelId" }
-  case class ChannelClosed(channelId: ByteVector32) extends ChannelCommandResponse { override def toString  = s"closed channel $channelId" }
+/** response to [[Command]] requests */
+sealed trait CommandResponse[+C <: Command]
+sealed trait CommandSuccess[+C <: Command] extends CommandResponse[C]
+sealed trait CommandFailure[+C <: Command, +T <: Throwable] extends CommandResponse[C] { def t: Throwable }
+
+/** generic responses */
+final case class RES_SUCCESS[+C <: Command](cmd: C, channelId: ByteVector32) extends CommandSuccess[C]
+final case class RES_FAILURE[+C <: Command, +T <: Throwable](cmd: C, t: T) extends CommandFailure[C, T]
+
+/**
+ * special case for [[CMD_ADD_HTLC]]
+ * note that for this command there is gonna be two response patterns:
+ * - either [[RES_ADD_FAILED]]
+ * - or [[RES_SUCCESS[CMD_ADD_HTLC]]] followed by [[RES_ADD_COMPLETED]] (possibly a while later)
+ */
+final case class RES_ADD_FAILED[+T <: Throwable](c: CMD_ADD_HTLC, t: T, channelUpdate: Option[ChannelUpdate]) extends CommandFailure[CMD_ADD_HTLC, T] { override def toString = s"cannot add htlc with origin=${c.origin} reason=${t.getMessage}" }
```

Comment:
It would be nice to be able to restrict to `+T <: ChannelException`, isn't is possible already?
Can't you simplify the signature like that:

```
final case class RES_ADD_FAILED(c: CMD_ADD_HTLC, t: ChannelException, channelUpdate: Option[ChannelUpdate]) extends CommandFailure[CMD_ADD_HTLC, ChannelException]
```

Reply:
It would be nice, but I can't right now, because we're relying on generic `Try` for commitment methods.  See https://github.com/ACINQ/eclair/pull/1520#discussion_r482796394. However we could move to a `scodec`-style custom error messages but a bit of work is needed.

Reply:
Done with a0ae9f81e.